contracts/colonyNetwork/ColonyNetworkExtensions.sol

```diff
@@ -0,0 +1,135 @@
+    require(resolvers[_extensionId][version] == address(0x0), "colony-network-extension-already-set");
+    require(
+      version == 1 || resolvers[_extensionId][version - 1] != address(0x0),
+      "colony-network-extension-bad-version"
+    );
+
+    resolvers[_extensionId][version] = _resolver;
+
+    emit ExtensionAdded(_extensionId, version);
+  }
+
+  function installExtension(bytes32 _extensionId, uint256 _version)
+    public
+    stoppable
+    calledByColony
+  {
+    require(resolvers[_extensionId][_version] != address(0x0), "colony-network-extension-bad-version");
+    require(installations[_extensionId][msg.sender] == address(0x0), "colony-network-extension-already-installed");
+
+    EtherRouter extension = new EtherRouter();
+    installations[_extensionId][msg.sender] = address(extension);
+
+    extension.setResolver(resolvers[_extensionId][_version]);
+    ColonyExtension(address(extension)).install(msg.sender);
+
+    emit ExtensionInstalled(_extensionId, msg.sender, _version);
+  }
+
+  function upgradeExtension(bytes32 _extensionId, uint256 _newVersion)
+    public
+    stoppable
+    calledByColony
+  {
+    require(installations[_extensionId][msg.sender] != address(0x0), "colony-network-extension-not-installed");
+
+    address payable extension = installations[_extensionId][msg.sender];
+    require(_newVersion == ColonyExtension(extension).version() + 1, "colony-network-extension-bad-increment");
+    require(resolvers[_extensionId][_newVersion] != address(0x0), "colony-network-extension-bad-version");
+
+    EtherRouter(extension).setResolver(resolvers[_extensionId][_newVersion]);
+    ColonyExtension(extension).finishUpgrade();
+    assert(ColonyExtension(extension).version() == _newVersion);
+
+    emit ExtensionUpgraded(_extensionId, msg.sender, _newVersion);
+  }
+
+  function deprecateExtension(bytes32 _extensionId, bool _deprecated)
+    public
+    stoppable
+    calledByColony
+  {
+    ColonyExtension(installations[_extensionId][msg.sender]).deprecate(_deprecated);
+
+    emit ExtensionDeprecated(_extensionId, msg.sender, _deprecated);
+  }
+
+  function uninstallExtension(bytes32 _extensionId)
+    public
+    stoppable
+    calledByColony
+  {
+    require(installations[_extensionId][msg.sender] != address(0x0), "colony-network-extension-not-installed");
+
+    ColonyExtension extension = ColonyExtension(installations[_extensionId][msg.sender]);
+    installations[_extensionId][msg.sender] = address(0x0);
+    extension.uninstall();
+
+    emit ExtensionUninstalled(_extensionId, msg.sender);
+  }
+
+  // Public view functions
+
+  function getExtensionResolver(bytes32 _extensionId, uint256 _version)
+    public
+    view
+    returns (address)
+  {
+    return resolvers[_extensionId][_version];
+  }
+
+  function getExtensionInstallation(bytes32 _extensionId, address _colony)
+    public
+    view
+    returns (address)
+  {
+    return installations[_extensionId][_colony];
```

Comment:
I'm going back and forth on whether this should return the version as well.

Reply:
I don't think it's necessary -- you can get the version from the extension itself.